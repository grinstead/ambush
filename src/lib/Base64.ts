import {
  BinaryArray,
  appendAscii,
  littleEndian,
  shiftReadIndex,
} from "./BinaryArray.ts";

const CHAR_PAD = 61; // '='

export function readBase64(
  array: BinaryArray,
  byteLength: number = array.byteLength
): Uint8Array {
  if (!byteLength) return new Uint8Array();

  const start = shiftReadIndex(array, byteLength);
  const encoded = array._bytes;

  let length = byteLength;
  let overhang = length & 3;

  // test if data is aligned, in which case it may have a pad

  if (overhang === 0 && encoded[start + length - 1] === CHAR_PAD) {
    length -= encoded[start + length - 2] === CHAR_PAD ? 2 : 1;
    overhang = length & 3;
  }

  // starts as the length of the output array
  let j = overhang
    ? (3 / 4) * (length - overhang) + overhang - 1
    : (3 / 4) * length;
  const output = new Uint8Array(j);

  let i = start + length;
  let a, b, c, d;

  if (overhang === 3) {
    c = decodeByte(encoded[i - 1]);
    b = decodeByte(encoded[i - 2]);
    a = decodeByte(encoded[(i -= 3)]);

    // a   b   c
    // |  /|  /|
    // 6+2 4+4 -
    output[j - 1] = ((b & 0x0f) << 4) | (c >> 2);
    output[(j -= 2)] = (a << 2) | (b >> 4);
  } else if (overhang === 2) {
    // a   b
    // |  /|
    // 6+2 -
    b = decodeByte(encoded[i - 1]);
    a = decodeByte(encoded[(i -= 2)]);

    output[--j] = (a << 2) | (b >> 4);
  }

  while (i > start) {
    d = decodeByte(encoded[i - 1]);
    c = decodeByte(encoded[i - 2]);
    b = decodeByte(encoded[i - 3]);
    a = decodeByte(encoded[(i -= 4)]);

    // a   b   c d
    // |  /|  /| |
    // 6+2 4+4 2+6
    output[j - 1] = ((c & 0x03) << 6) | d;
    output[j - 2] = ((b & 0x0f) << 4) | (c >> 2);
    output[(j -= 3)] = (a << 2) | (b >> 4);
  }

  return output;
}

function decodeByte(byte: number) {
  return byte === 43 /* '+' */
    ? 62
    : byte === 47 /* '/' */
    ? 63
    : byte <= 57 /* '9' */
    ? byte + 4 // 61 - ('9' - byte)
    : byte <= 90 /* 'Z' */
    ? byte - 65 // 25 - ('Z' - byte)
    : byte - 71; // 51 - ('z' - byte)
}

export function parseBase64(base64: string) {
  const bin = littleEndian();
  appendAscii(bin, base64);
  return readBase64(bin);
}
