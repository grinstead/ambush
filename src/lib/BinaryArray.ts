export const NUM_BYTES_INT16 = 2;
export const NUM_BYTES_INT32 = 4;
export const NUM_BYTES_FLOAT32 = 4;

// arbitrarily chosen
const START_SIZE = 32;

/**
 * A class that helps parse and write binary data.
 *
 * It contains a read and write index. When you read from the DataArray, it will
 * shift the read index, so that reading an int32 and then reading an int32
 * again is equivalent to taking the front 8 bytes and interpreting it as two
 * ints.
 *
 * Writing to this class is done almost exclusively with appends. The class
 * will automatically resize its internal buffer if you write past the end of
 * what is currently allocated.
 */
export class BinaryArray {
  _bytes: Uint8Array;
  _view: undefined | DataView;
  _readIndex: number = 0;
  _writeIndex: number;
  _allocated: number;
  littleEndian: boolean = false;

  /**
   * @param bytes The bytes to read from
   * @param overwrite If this is true, the data will instead write to this buffer
   */
  constructor(
    bytes?: ArrayBuffer | ArrayBufferView,
    overwrite: boolean = !bytes
  ) {
    let asUint8;
    if (bytes instanceof ArrayBuffer) {
      asUint8 = new Uint8Array(bytes);
    } else if (bytes) {
      asUint8 =
        bytes instanceof Uint8Array
          ? bytes
          : new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    } else {
      asUint8 = new Uint8Array(START_SIZE);
    }

    this._bytes = asUint8;
    this._allocated = asUint8.length;
    this._writeIndex = overwrite ? 0 : asUint8.byteLength;
  }

  get byteLength() {
    return this._writeIndex - this._readIndex;
  }
}

/**
 * A Little-Endian Binary Array
 */
export type BinaryArrayLE = BinaryArray & { littleEndian: true };
/**
 * A Big-Endian Binary Array
 */
export type BinaryArrayBE = BinaryArray & { littleEndian: false };

export function asUint8Array(array: BinaryArray): Uint8Array {
  const { _readIndex: start, _writeIndex: end, _bytes: bytes } = array;

  return start < end
    ? new Uint8Array(bytes.buffer, bytes.byteOffset + start, end - start)
    : new Uint8Array();
}

export function bigEndian(
  bytes?: ArrayBuffer | ArrayBufferView,
  overwrite?: boolean
): BinaryArrayBE {
  return new BinaryArray(bytes, overwrite) as BinaryArrayBE;
}

export function littleEndian(
  bytes?: ArrayBuffer | ArrayBufferView,
  overwrite?: boolean
): BinaryArrayLE {
  const array = new BinaryArray(bytes, overwrite) as BinaryArrayLE;
  array.littleEndian = true;
  return array;
}

export function numBytes(array: BinaryArray) {
  return array._writeIndex;
}

function v(array: BinaryArray): DataView {
  let view = array._view;
  if (!view) {
    const bytes = array._bytes;
    array._view = view = new DataView(
      bytes.buffer,
      bytes.byteOffset,
      bytes.byteLength
    );
  }

  return view;
}

export { v as viewOf };

export function ensureAdditionalBytes(array: BinaryArray, numBytes: number) {
  // "claim" the bytes but then reset it back down
  array._writeIndex = claimBytes(array, numBytes);
}

export function shiftReadIndex(array: BinaryArray, count: number): number {
  const start = array._readIndex;
  const shifted = start + count;

  shifted > array._writeIndex && throwBinaryArrayError(1);

  array._readIndex = shifted;
  return start;
}

export function claimBytes(array: BinaryArray, numBytes: number): number {
  let { _writeIndex: index, _allocated: allocated } = array;

  if ((array._writeIndex = index + numBytes) > allocated) {
    const offset = array._readIndex;

    // we have to re-allocate the underlying buffer, either double the size, or
    // add some bytes if the bytesize is tiny
    allocated = Math.max(2 * (allocated - offset), START_SIZE, numBytes);

    const prev = array._bytes;
    const data = new Uint8Array(allocated);
    data.set(offset ? prev.subarray(offset) : prev);

    array._readIndex = 0;
    array._writeIndex = (index -= offset) + numBytes;
    array._allocated = allocated;
    array._bytes = data;
    array._view = undefined;
  }

  return index;
}

export function appendUint8(array: BinaryArray, value: number) {
  const index = claimBytes(array, 1);
  array._bytes[index] = value;
}

export function appendUint16(
  array: BinaryArray,
  value: number,
  littleEndian: boolean = array.littleEndian
) {
  const index = claimBytes(array, NUM_BYTES_INT16);
  v(array).setUint16(index, value, littleEndian);
}

export function appendUint32(
  array: BinaryArray,
  value: number,
  littleEndian: boolean = array.littleEndian
) {
  const index = claimBytes(array, NUM_BYTES_INT32);
  v(array).setUint32(index, value, littleEndian);
}

export function appendFloat32(
  array: BinaryArray,
  value: number,
  littleEndian: boolean = array.littleEndian
) {
  const index = claimBytes(array, NUM_BYTES_FLOAT32);
  v(array).setFloat32(index, value, littleEndian);
}

export function appendUint8Array(array: BinaryArray, buffer: Uint8Array) {
  const index = claimBytes(array, buffer.byteLength);
  array._bytes.set(buffer, index);
}

let utf8Encoder: undefined | TextEncoder;
export function appendUtf8(
  array: BinaryArray,
  text: string,
  maxBytes?: number
): TextEncoderEncodeIntoResult {
  // an over-allocation, but the text will be guaranteed to fit
  const length = maxBytes ?? text.length * 3;
  const index = claimBytes(array, length);

  const bytes = array._bytes.subarray(index, index + length);
  const result = (utf8Encoder ??= new TextEncoder()).encodeInto(text, bytes);

  // reset the write index to reclaim any potentially unused bytes
  array._writeIndex = index + result.written;

  return result;
}

export function readUint8(array: BinaryArray): number {
  const index = shiftReadIndex(array, 1);
  return array._bytes[index];
}

export function readUint16(
  array: BinaryArray,
  littleEndian: boolean = array.littleEndian
): number {
  const index = shiftReadIndex(array, 2);
  return v(array).getUint16(index, littleEndian);
}

export function readUint32(
  array: BinaryArray,
  littleEndian: boolean = array.littleEndian
): number {
  const index = shiftReadIndex(array, 4);
  return v(array).getUint32(index, littleEndian);
}

export function readFloat32(
  array: BinaryArray,
  littleEndian: boolean = array.littleEndian
): number {
  const index = shiftReadIndex(array, 4);
  return v(array).getFloat32(index, littleEndian);
}

export function readUint8Array(array: BinaryArray, length: number): Uint8Array {
  const index = shiftReadIndex(array, length);
  return array._bytes.subarray(index, index + length);
}

let utf8Decoder: undefined | TextDecoder;
export function readUtf8(array: BinaryArray, length: number): string {
  const index = shiftReadIndex(array, length);

  return (utf8Decoder ??= new TextDecoder()).decode(
    array._bytes.subarray(index, index + length)
  );
}

function throwBinaryArrayError(errorCode: number) {
  throw new Error(`BinaryArrayError (code ${errorCode})`);
}
