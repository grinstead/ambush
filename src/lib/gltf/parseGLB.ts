import {
  BinaryArray,
  BinaryArrayLE,
  NUM_BYTES_INT32,
  littleEndian,
  readUtf8,
  readUint32,
  readUint8,
  readUint8Array,
} from "../BinaryArray.ts";
import { GLTFAsset } from "./gltf_types.ts";

const HEADER_MAGIC = 0x46546c67; // glTF in ASCII (when little-endian), all glb files begin with this
const SUPPORTED_VERSION = 2;
const HEADER_LENGTH = 3 * NUM_BYTES_INT32;
// glb files MUST contain a json chunk
const MIN_LENGTH = HEADER_LENGTH + 2 * NUM_BYTES_INT32 + 2; // 2 === '{}'.length;
const LITTLE_ENDIAN = true;
const CHUNK_HEADER_JSON = 0x4e4f534a;
const CHUNK_HEADER_BIN = 0x004e4942;

export type GLBFile = {
  asset: GLTFAsset;
  bin?: Uint8Array;
};

export function parseGLB(container: BinaryArray): GLBFile {
  const binary = claimGLBContents(container);

  const jsonLength = readUint32(binary);
  const jsonType = readUint32(binary);

  if (jsonType !== CHUNK_HEADER_JSON) {
    throw new Error(`Unrecognized Initial Chunk ${jsonType}`);
  }

  const unparsedJson = readUtf8(binary, jsonLength);

  let bin: undefined | Uint8Array;

  // the spec guarantees if there is a binary chunk, that it will be the second
  // chunk. Additionally, we should ignore any unrecognized chunk types
  if (binary.byteLength > 0) {
    const binLength = readUint32(binary);
    const binType = readUint32(binary);

    if (binType === CHUNK_HEADER_BIN) {
      bin = readUint8Array(binary, binLength);
    }
  }

  // for true correctness, we should check that all remaining data is structured
  // into the spec's chunk format, but why bother.

  // TODO: we should validate this
  const asset = JSON.parse(unparsedJson) as GLTFAsset;

  return { asset, bin };
}

function claimGLBContents(binary: BinaryArray): BinaryArrayLE {
  const byteLength = binary.byteLength;

  if (byteLength < MIN_LENGTH) {
    throw new Error(
      `Given impossibly small (${binary.byteLength} bytes) glb file`
    );
  }

  if (readUint32(binary, LITTLE_ENDIAN) !== HEADER_MAGIC) {
    throw new Error("Given non-glb data");
  }

  const version = readUint32(binary, LITTLE_ENDIAN);
  if (version !== SUPPORTED_VERSION) {
    throw new Error(`Unrecognized .glb version ${version}`);
  }

  const length = readUint32(binary, LITTLE_ENDIAN);
  if (length > byteLength) {
    throw new Error(
      `Incomplete .glb file (expected ${length} bytes, given ${byteLength})`
    );
  }

  return littleEndian(readUint8Array(binary, length - HEADER_LENGTH));
}
