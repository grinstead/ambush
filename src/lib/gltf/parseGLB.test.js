import { expect, test } from "bun:test";
import { readBase64 } from "../Base64.ts";
import { parseGLB } from "./parseGLB.ts";
import { appendUint8Array, appendUtf8, littleEndian } from "../BinaryArray.ts";

// https://github.com/KhronosGroup/glTF-Sample-Assets/blob/main/Models/Box/glTF-Binary/Box.glb
const BOX_FILE =
  "Z2xURgIAAACABgAA3AMAAEpTT057ImFzc2V0Ijp7ImdlbmVyYXRvciI6IkNPTExBREEyR0xURiIsInZlcnNpb24iOiIyLjAifSwic2NlbmUiOjAsInNjZW5lcyI6W3sibm9kZXMiOlswXX1dLCJub2RlcyI6W3siY2hpbGRyZW4iOlsxXSwibWF0cml4IjpbMS4wLDAuMCwwLjAsMC4wLDAuMCwwLjAsLTEuMCwwLjAsMC4wLDEuMCwwLjAsMC4wLDAuMCwwLjAsMC4wLDEuMF19LHsibWVzaCI6MH1dLCJtZXNoZXMiOlt7InByaW1pdGl2ZXMiOlt7ImF0dHJpYnV0ZXMiOnsiTk9STUFMIjoxLCJQT1NJVElPTiI6Mn0sImluZGljZXMiOjAsIm1vZGUiOjQsIm1hdGVyaWFsIjowfV0sIm5hbWUiOiJNZXNoIn1dLCJhY2Nlc3NvcnMiOlt7ImJ1ZmZlclZpZXciOjAsImJ5dGVPZmZzZXQiOjAsImNvbXBvbmVudFR5cGUiOjUxMjMsImNvdW50IjozNiwibWF4IjpbMjNdLCJtaW4iOlswXSwidHlwZSI6IlNDQUxBUiJ9LHsiYnVmZmVyVmlldyI6MSwiYnl0ZU9mZnNldCI6MCwiY29tcG9uZW50VHlwZSI6NTEyNiwiY291bnQiOjI0LCJtYXgiOlsxLjAsMS4wLDEuMF0sIm1pbiI6Wy0xLjAsLTEuMCwtMS4wXSwidHlwZSI6IlZFQzMifSx7ImJ1ZmZlclZpZXciOjEsImJ5dGVPZmZzZXQiOjI4OCwiY29tcG9uZW50VHlwZSI6NTEyNiwiY291bnQiOjI0LCJtYXgiOlswLjUsMC41LDAuNV0sIm1pbiI6Wy0wLjUsLTAuNSwtMC41XSwidHlwZSI6IlZFQzMifV0sIm1hdGVyaWFscyI6W3sicGJyTWV0YWxsaWNSb3VnaG5lc3MiOnsiYmFzZUNvbG9yRmFjdG9yIjpbMC44MDAwMDAwMTE5MjA5MjksMC4wLDAuMCwxLjBdLCJtZXRhbGxpY0ZhY3RvciI6MC4wfSwibmFtZSI6IlJlZCJ9XSwiYnVmZmVyVmlld3MiOlt7ImJ1ZmZlciI6MCwiYnl0ZU9mZnNldCI6NTc2LCJieXRlTGVuZ3RoIjo3MiwidGFyZ2V0IjozNDk2M30seyJidWZmZXIiOjAsImJ5dGVPZmZzZXQiOjAsImJ5dGVMZW5ndGgiOjU3NiwiYnl0ZVN0cmlkZSI6MTIsInRhcmdldCI6MzQ5NjJ9XSwiYnVmZmVycyI6W3siYnl0ZUxlbmd0aCI6NjQ4fV19iAIAAEJJTgAAAAAAAAAAAAAAgD8AAAAAAAAAAAAAgD8AAAAAAAAAAAAAgD8AAAAAAAAAAAAAgD8AAAAAAACAvwAAAAAAAAAAAACAvwAAAAAAAAAAAACAvwAAAAAAAAAAAACAvwAAAAAAAIA/AAAAAAAAAAAAAIA/AAAAAAAAAAAAAIA/AAAAAAAAAAAAAIA/AAAAAAAAAAAAAAAAAACAPwAAAAAAAAAAAACAPwAAAAAAAAAAAACAPwAAAAAAAAAAAACAPwAAAAAAAIC/AAAAAAAAAAAAAIC/AAAAAAAAAAAAAIC/AAAAAAAAAAAAAIC/AAAAAAAAAAAAAAAAAAAAAAAAgL8AAAAAAAAAAAAAgL8AAAAAAAAAAAAAgL8AAAAAAAAAAAAAgL8AAAC/AAAAvwAAAD8AAAA/AAAAvwAAAD8AAAC/AAAAPwAAAD8AAAA/AAAAPwAAAD8AAAA/AAAAvwAAAD8AAAC/AAAAvwAAAD8AAAA/AAAAvwAAAL8AAAC/AAAAvwAAAL8AAAA/AAAAPwAAAD8AAAA/AAAAvwAAAD8AAAA/AAAAPwAAAL8AAAA/AAAAvwAAAL8AAAC/AAAAPwAAAD8AAAA/AAAAPwAAAD8AAAC/AAAAPwAAAL8AAAA/AAAAPwAAAL8AAAC/AAAAvwAAAD8AAAC/AAAAPwAAAD8AAAC/AAAAvwAAAL8AAAC/AAAAPwAAAL8AAAC/AAAAvwAAAL8AAAC/AAAAPwAAAL8AAAA/AAAAvwAAAL8AAAA/AAAAPwAAAL8AAAEAAgADAAIAAQAEAAUABgAHAAYABQAIAAkACgALAAoACQAMAA0ADgAPAA4ADQAQABEAEgATABIAEQAUABUAFgAXABYAFQA=";

test("Box file", () => {
  const binary = littleEndian();
  appendUtf8(binary, BOX_FILE);
  appendUint8Array(binary, readBase64(binary));

  expect(parseGLB(binary).asset).toMatchSnapshot();
});
