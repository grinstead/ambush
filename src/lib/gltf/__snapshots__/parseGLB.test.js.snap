// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Box file 1`] = `
{
  "accessors": [
    {
      "bufferView": 0,
      "byteOffset": 0,
      "componentType": 5123,
      "count": 36,
      "max": [
        23,
      ],
      "min": [
        0,
      ],
      "type": "SCALAR",
    },
    {
      "bufferView": 1,
      "byteOffset": 0,
      "componentType": 5126,
      "count": 24,
      "max": [
        1,
        1,
        1,
      ],
      "min": [
        -1,
        -1,
        -1,
      ],
      "type": "VEC3",
    },
    {
      "bufferView": 1,
      "byteOffset": 288,
      "componentType": 5126,
      "count": 24,
      "max": [
        0.5,
        0.5,
        0.5,
      ],
      "min": [
        -0.5,
        -0.5,
        -0.5,
      ],
      "type": "VEC3",
    },
  ],
  "asset": {
    "generator": "COLLADA2GLTF",
    "version": "2.0",
  },
  "bufferViews": [
    {
      "buffer": 0,
      "byteLength": 72,
      "byteOffset": 576,
      "target": 34963,
    },
    {
      "buffer": 0,
      "byteLength": 576,
      "byteOffset": 0,
      "byteStride": 12,
      "target": 34962,
    },
  ],
  "buffers": [
    {
      "byteLength": 648,
    },
  ],
  "materials": [
    {
      "name": "Red",
      "pbrMetallicRoughness": {
        "baseColorFactor": [
          0.800000011920929,
          0,
          0,
          1,
        ],
        "metallicFactor": 0,
      },
    },
  ],
  "meshes": [
    {
      "name": "Mesh",
      "primitives": [
        {
          "attributes": {
            "NORMAL": 1,
            "POSITION": 2,
          },
          "indices": 0,
          "material": 0,
          "mode": 4,
        },
      ],
    },
  ],
  "nodes": [
    {
      "children": [
        1,
      ],
      "matrix": [
        1,
        0,
        0,
        0,
        0,
        0,
        -1,
        0,
        0,
        1,
        0,
        0,
        0,
        0,
        0,
        1,
      ],
    },
    {
      "mesh": 0,
    },
  ],
  "scene": 0,
  "scenes": [
    {
      "nodes": [
        0,
      ],
    },
  ],
}
`;
